File: first_non_repeating_character/first_non_repeating_character.js
/**
 * Write a function that takes a string as input and returns the first non-repeating character in the string. If there is no non-repeating character, return null.
 * For example:
 * For the input "aabccd", the function should return 'b'.
 * For the input "aabbcc", the function should return null.
 *
 * This question is good because:
 * It requires the use of a hash table (object) for efficient counting.
 * It involves string manipulation and iteration.
 *
 * @param {string} str - The input string
 * @return {string|null} The first non-repeating character, or null if none exists
 */

function firstSolution(str) {
  const charCount = {};
  for (let i = 0; i < str.length; i++) {
    const char = str[i];
    if (charCount[char]) {
      charCount[char]++;
    } else {
      charCount[char] = 1;
    }
  }

  for (let i = 0; i < str.length; i++) {
    if (charCount[str[i]] === 1) {
      return str[i];
    }
  }
  return null;
}

function second_solution(str) {
  // uses modern JS syntax object property access
  const charMap = new Map();

  for (const char of str) {
    charMap.set(char, (charMap.get(char) || 0) + 1);
  }

  for (const [char, count] of charMap) {
    if (count === 1) {
      return char;
    }
  }

  return null;
}

function firstNonRepeatingCharacter(str) {
  //   return firstSolution(str);
  return second_solution(str);
}

// Test cases
console.log(firstNonRepeatingCharacter("aabccd")); // Should output: "b"
console.log(firstNonRepeatingCharacter("aabbcc")); // Should output: null
console.log(firstNonRepeatingCharacter("abcdef")); // Should output: "a"
console.log(firstNonRepeatingCharacter("aabbccdeeff")); // Should output: "d"

File: first_non_repeating_character/first_non_repeating_character.rb
# find the first non repeating character in a string

def first_hash_soluti0on(str)
  # most efficient. O(n)
  # best overall solution
  char_map = Hash.new(0)
  
  str.each_char do |char|
    char_map[char] += 1
  end
  
  str.each_char do |char|
    return char if char_map[char] == 1
  end
  
  'nothing'
end

def second_solution_using_array(str)
  # most concise but less efficient. O(n^2)
  str.chars.find { |char| str.count(char) == 1 } || 'nothing'
end

def third_solution_using_ruby_tally(str)
  # clean and readable. O(n)
  # use if flexing ruby knowledge. Idiomatic ruby example
  chars = str.chars
  char_counts = chars.tally
  chars.find { |char| char_counts[char] == 1 } || 'nothing'
end

def fourth_solution_using_ruby_each_with_index(str)
  # more verbose but less effeciant. O(n^2)
  chars = str.chars
  chars.each_with_index do |char, index|
    return char if chars.count(char) == 1
  end
  'nothing'
end

def first_non_repeating_character(str)
  # first_hash_soluti0on(str)
  # second_solution_using_array(str)
  # third_solution_using_ruby_tally(str)
  fourth_solution_using_ruby_each_with_index(str)
end

# Test cases
puts first_non_repeating_character("aabccd")    # Should output: "b"
puts first_non_repeating_character("aabbcc")    # Should output: "nothing"
puts first_non_repeating_character("abcdef")    # Should output: "a"
puts first_non_repeating_character("aabbccdeeff")  # Should output: "d"